results/pkg/watcher/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -0,0 +1,137 @@
+package pipelinerun
+
+import (
+	"context"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/experimental/results/pkg/watcher/convert"
+	"github.com/tektoncd/experimental/results/pkg/watcher/reconciler/common"
+	pb "github.com/tektoncd/experimental/results/proto/proto"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	ttesting "github.com/tektoncd/pipeline/pkg/reconciler/testing"
+	"github.com/tektoncd/pipeline/test"
+	"google.golang.org/protobuf/testing/protocmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/configmap"
+)
+
+func TestReconcile(t *testing.T) {
+	pipelineRunTest := NewPipelineRunTest(t)
+
+	testFuncs := map[string]func(t *testing.T){
+		"Create":   pipelineRunTest.testCreatePipelineRun,
+		"Unchange": pipelineRunTest.testUnchangePipelineRun,
+		"Update":   pipelineRunTest.testUpdatePipelineRun,
+	}
+
+	for name, testFunc := range testFuncs {
+		t.Run(name, testFunc)
+	}
+}
+
+type PipelineRunTest struct {
+	pipelineRun *v1beta1.PipelineRun
+	asset       test.Assets
+	ctx         context.Context
+	client      pb.ResultsClient
+}
+
+func NewPipelineRunTest(t *testing.T) PipelineRunTest {
+	client := common.NewResultsClient(t)
+	pipelineRun := &v1beta1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "Tekton-PipelineRun",
+			Namespace:   "default",
+			Annotations: map[string]string{"demo": "pipelinerun_demo"},
+			UID:         "54321",
+		},
+	}
+	asset, ctx := getFakeClients(t, []*v1beta1.PipelineRun{pipelineRun}, client)
+	pipelineRunTest := PipelineRunTest{
+		pipelineRun: pipelineRun,
+		asset:       asset,
+		ctx:         ctx,
+		client:      client,
+	}
+	return pipelineRunTest
+}
+
+func (tt *PipelineRunTest) testCreatePipelineRun(t *testing.T) {
+	pr, err := common.ReconcilePipelineRun(tt.ctx, tt.asset, tt.pipelineRun)
+	if err != nil {
+		t.Fatalf("Failed to get completed PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	if _, ok := pr.Annotations[common.IDName]; !ok {
+		t.Fatalf("Expected completed PipelineRun %s should be updated with a results_id field in annotations", tt.pipelineRun.Name)
+	}
+	if _, err := tt.client.GetResult(tt.ctx, &pb.GetResultRequest{Name: pr.Annotations[common.IDName]}); err != nil {
+		t.Fatalf("Expected completed PipelineRun %s not created in api server", tt.pipelineRun.Name)
+	}
+}
+
+func (tt *PipelineRunTest) testUnchangePipelineRun(t *testing.T) {
+	pr, err := tt.asset.Clients.Pipeline.TektonV1beta1().PipelineRuns(tt.pipelineRun.Namespace).Get(tt.pipelineRun.Name, metav1.GetOptions{})
```

Comment:
What's the difference between pr and tt.pipelineRun?

Reply:
`tt.pipelineRun` is the seed test data, while `pr` being the newest version from the client. There might be some change in the remote client.